include/aspect/material_model/cpp.h

```diff
@@ -0,0 +1,116 @@
+/*
+  Copyright (C) 2018 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef _aspect_material_model_cpp_h
+#define _aspect_material_model_cpp_h
+
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+#include <aspect/utilities.h>
+#include <regex>
+
+#if ASPECT_USE_SHARED_LIBS==1
+#  include <dlfcn.h>
+#  ifdef ASPECT_HAVE_LINK_H
+#    include <link.h>
+#  endif
+#endif
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    /**
+     * A material model that compiles user-defined C++
+     * code at runtime.
+     *
+     * @ingroup MaterialModels
+     */
+    template <int dim>
+    class CPP : public MaterialModel::Interface<dim>, public ::aspect::SimulatorAccess<dim>
```

Comment:
My thought was that the class name doesn't matter so much, but it might be worthwhile making the class name, the model name, and the parameter subsection consistent with one another. Right now they're `CPP`, `cpp`, and `C plus plus`, respectively.